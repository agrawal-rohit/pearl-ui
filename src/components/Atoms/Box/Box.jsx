import React from "react";
import { View } from "react-native";
import useStyledProps from "../../../hooks/useStyledProps";
import { useTheme } from "../../../hooks/useTheme";
import {
  backgroundColor,
  border,
  layout,
  opacity,
  position,
  shadow,
  spacing,
  visible,
} from "../../../theme/src/styleFunctions";
import {
  IBasePearlTheme,
  RestyleFunctionContainer,
} from "../../../theme/src/types";

// export type BoxProps<
//   Theme extends IBasePearlTheme
// > = BackgroundColorProps<Theme> &
//   OpacityProps &
//   VisibleProps &
//   LayoutProps &
//   SpacingProps<Theme> &
//   BorderProps<Theme> &
//   ShadowProps<Theme> &
//   PositionProps<Theme>;

export const boxRestyleFunctions = [
  backgroundColor,
  opacity,
  visible,
  layout,
  spacing,
  border,
  shadow,
  position,
];

/**
 * Box is the most abstract component on top of which all other Pearl UI components are built. By default, it renders a <View> element.
 */
// type ViewProps = React.ComponentProps<typeof View> & {
//   children?: React.ReactNode;
// };
// type ComponentProps<Theme extends IBasePearlTheme> = BoxProps<Theme> &
//   Omit<ViewProps, keyof BoxProps<Theme>>;

const Box = React.forwardRef((props, ref) => {
  const passedProps = useStyledProps(boxRestyleFunctions, props);

  return <View ref={ref} {...passedProps} />;
});

export default Box;
