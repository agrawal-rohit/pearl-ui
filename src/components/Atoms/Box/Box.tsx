import React from "react";
import { View, ViewProps } from "react-native";
import { pearlify } from "../../../pearlify";

/**
 * Box is the most abstract component on top of which all other Pearl UI components are built.
 * By default, it renders a <View> element.
 */
const Box = pearlify<ViewProps>(View, {
  componentName: "Box",
  type: "basic",
  animatable: true,
});

// The props that the Box component accepts.
export type BoxProps = React.ComponentProps<typeof Box>;

// Export the Box component as the default export.
export default Box;
