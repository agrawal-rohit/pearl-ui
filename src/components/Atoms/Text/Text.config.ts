export default {
  baseStyle: {
    color: {
      light: "neutral.900",
      dark: "neutral.50",
    },
    scaleFontSize: true,
  },
  variants: {
    h1: {
      fontFamily: "heading",
      fontWeight: "bold",
      fontSize: "8xl",
      lineHeight: "10xl",
    },
    h2: {
      fontFamily: "heading",
      fontWeight: "bold",
      fontSize: "5xl",
      lineHeight: "7xl",
    },
    t1: {
      fontFamily: "heading",
      fontWeight: "semibold",
      fontSize: "2xl",
      lineHeight: "4xl",
    },
    t2: {
      fontFamily: "heading",
      fontWeight: "semibold",
      fontSize: "l",
      lineHeight: "2xl",
    },
    st1: {
      fontFamily: "heading",
      fontWeight: "semibold",
      fontSize: "m",
      lineHeight: "xl",
    },
    st2: {
      fontFamily: "heading",
      fontWeight: "semibold",
      fontSize: "s",
      lineHeight: "l",
    },
    p1: {
      fontFamily: "body",
      fontWeight: "normal",
      fontSize: "m",
      lineHeight: "xl",
    },
    p2: {
      fontFamily: "body",
      fontWeight: "normal",
      fontSize: "s",
      lineHeight: "l",
    },
    caption: {
      fontFamily: "body",
      fontWeight: "normal",
      fontSize: "xs",
      lineHeight: "m",
    },
    btn1: {
      fontFamily: "body",
      fontWeight: "medium",
      fontSize: "m",
      lineHeight: "xl",
    },
    btn2: {
      fontFamily: "body",
      fontWeight: "medium",
      fontSize: "s",
      lineHeight: "m",
    },
    btn3: {
      fontFamily: "body",
      fontWeight: "medium",
      fontSize: "xs",
      lineHeight: "s",
    },
    btn4: {
      fontFamily: "body",
      fontWeight: "medium",
      fontSize: "2xs",
      lineHeight: "xs",
    },
  },
  defaults: {
    variant: "p1",
  },
};
