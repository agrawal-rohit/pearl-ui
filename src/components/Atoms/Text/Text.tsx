import React from "react";
import { Text as RNText } from "react-native";
import responsiveSize from "../../../utils/responsiveSize";
import { useAtomicComponentConfig } from "../../../hooks/useAtomicComponentConfig";
import {
  backgroundColor,
  BackgroundColorProps,
  color,
  ColorProps,
  layout,
  LayoutProps,
  opacity,
  OpacityProps,
  spacing,
  SpacingProps,
  textShadow,
  TextShadowProps,
  typography,
  TypographyProps,
  visible,
  VisibleProps,
} from "../../../theme/src/styleFunctions";
import {
  ComponentSizes,
  ComponentVariants,
  ResponsiveValue,
  StyleFunctionContainer,
} from "../../../theme/src/types";
import { useTheme } from "../../../hooks/useTheme";

type TextStyleProps = ColorProps &
  BackgroundColorProps &
  OpacityProps &
  VisibleProps &
  LayoutProps &
  TypographyProps &
  SpacingProps &
  TextShadowProps & {
    /** The size of the text */
    size?: ResponsiveValue<ComponentSizes<"Text">>;
    /** The variant of the text */
    variant?: ResponsiveValue<ComponentVariants<"Text">>;
    /** Whether to slightly scale the font size based on the screen dimensions */
    scaleFontSize?: boolean;
  };

export const textStyleFunctions = [
  color,
  backgroundColor,
  opacity,
  visible,
  layout,
  typography,
  spacing,
  textShadow,
] as StyleFunctionContainer[];

type RNTextProps = React.ComponentProps<typeof RNText>;

export type TextProps = TextStyleProps &
  Omit<RNTextProps, keyof TextStyleProps>;

export const buildFontConfig = (
  textStyle: any,
  allowFontScaling: boolean
): object => {
  const fontWeight = textStyle.fontWeight;
  const fontStyle = textStyle.fontStyle || "normal";

  let fontSize = textStyle.fontSize;
  let lineHeight = textStyle.lineHeight;

  if (allowFontScaling) {
    fontSize = responsiveSize(fontSize);
    lineHeight = responsiveSize(lineHeight);
  }

  const initialFontFamily = textStyle.fontFamily;
  const { theme } = useTheme();

  const finalFontFamily =
    theme.fontConfig[initialFontFamily][fontWeight][fontStyle];

  return {
    fontSize,
    fontWeight,
    fontFamily: finalFontFamily,
    fontStyle,
  };
};

/**
 * Text is the component which controls the typography across your app. By default, it renders a <Text /> component
 */
const Text = React.forwardRef((rest: TextProps, ref: any) => {
  const props = useAtomicComponentConfig(
    "Text",
    rest,
    {
      size: rest.size,
      variant: rest.variant,
    },
    "primary",
    textStyleFunctions
  );

  const memoizedBuildFontConfig = React.useCallback(
    () => buildFontConfig(props.style, props.scaleFontSize),
    [props.style, props.scaleFontSize]
  );

  props.style = {
    includeFontPadding: false,
    ...props.style,
    ...memoizedBuildFontConfig(),
  };

  return (
    <RNText
      ref={ref}
      accessible={true}
      accessibilityRole="text"
      allowFontScaling
      {...props}
    >
      {props.children}
    </RNText>
  );
});

export default Text;
