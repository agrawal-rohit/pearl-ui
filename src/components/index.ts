// Atoms
export { default as Box, BoxProps } from "./atoms/box/box";
export { default as Center } from "./atoms/center/center";
export { default as Collapse, CollapseProps } from "./atoms/collapse/collapse";
export { default as Divider, DividerProps } from "./atoms/divider/divider";
export { default as Fade, FadeProps } from "./atoms/fade/fade";
export { default as Grid, GridProps } from "./atoms/grid/grid";
export { default as Icon, IconProps } from "./atoms/icon/icon";
export {
  default as Pressable,
  PressableProps,
} from "./atoms/pressable/pressable";
export {
  default as ScaleFade,
  ScaleFadeProps,
} from "./atoms/scale-fade/scale-fade";
export {
  default as List,
  OrderedList,
  UnorderedList,
  ListProps,
} from "./atoms/list/list";
export { default as Screen, ScreenProps } from "./atoms/screen/screen";
export { default as Skeleton, SkeletonProps } from "./atoms/skeleton/skeleton";
export {
  default as SkeletonCircle,
  SkeletonCircleProps,
} from "./atoms/skeleton/skeleton-circle";
export {
  default as SkeletonText,
  SkeletonTextProps,
} from "./atoms/skeleton/skeleton-text";
export { default as Slide, SlideProps } from "./atoms/slide/slide";
export {
  default as SlideFade,
  SlideFadeProps,
} from "./atoms/slide-fade/slide-fade";
export { default as Spacer } from "./atoms/spacer/spacer";
export { default as Spinner, SpinnerProps } from "./atoms/spinner/spinner";
export {
  default as Stack,
  HStack,
  VStack,
  ZStack,
  HStackProps,
  VStackProps,
  StackProps,
} from "./atoms/stack/stack";
export { default as Text, TextProps } from "./atoms/text/text";

// Molecules
export { default as Avatar, AvatarProps } from "./molecules/avatar/avatar";
export {
  default as AvatarGroup,
  AvatarGroupProps,
} from "./molecules/avatar/avatar-group";
export { default as Badge, BadgeProps } from "./molecules/badge/badge";
export {
  default as withBadge,
  WithBadgeOptions,
} from "./molecules/badge/withBadge";
export { default as Button, ButtonProps } from "./molecules/button/button";
export {
  default as ButtonGroup,
  ButtonGroupProps,
} from "./molecules/button/button-group";
export {
  default as CheckBox,
  CheckBoxProps,
} from "./molecules/checkbox/checkbox";
export {
  default as CheckBoxGroup,
  CheckBoxGroupProps,
} from "./molecules/checkbox/checkbox-group";
export {
  default as IconButton,
  IconButtonProps,
} from "./molecules/icon-button/icon-button";
export { default as Image, ImageProps } from "./molecules/image/image";
export { default as Input, InputProps } from "./molecules/input/input";
export {
  default as PinInput,
  PinInputProps,
} from "./molecules/pin-input/pin-input";
export {
  default as Progress,
  ProgressProps,
} from "./molecules/progress/progress";
export { default as Radio, RadioProps } from "./molecules/radio/radio";
export {
  default as RadioGroup,
  RadioGroupProps,
} from "./molecules/radio/radio-group";
export { default as Switch, SwitchProps } from "./molecules/switch/switch";
export {
  default as Textarea,
  TextareaProps,
} from "./molecules/textarea/textarea";
export { default as Video, VideoProps } from "./molecules/video/video";
export {
  default as Accordion,
  AccordionProps,
  AccordionItemProps,
  AccordionButtonProps,
  AccordionPanelProps,
  AccordionItem,
  AccordionButton,
  AccordionPanel,
} from "./molecules/accordion/accordion";
export { default as Alert, AlertProps } from "./molecules/alert/alert";
