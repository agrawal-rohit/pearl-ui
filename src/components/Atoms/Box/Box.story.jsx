import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "../Screen/Screen";
import Box from "./Box";

storiesOf("Box", module)
  .addDecorator((getStory) => getStory())
  .add("Basic box", () => (
    <Box
      padding="s"
      height={300}
      width={100}
      backgroundColor="primary100"
    ></Box>
  ));
