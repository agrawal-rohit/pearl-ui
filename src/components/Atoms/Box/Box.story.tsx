import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "../Screen/Screen";
import Box from "../Box/Box";

storiesOf("Box", module)
  .addDecorator((getStory) => <Screen>{getStory()}</Screen>)
  .add("Basic box", () => (
    <Box padding="s" width={200} backgroundColor="primary100"></Box>
  ));
