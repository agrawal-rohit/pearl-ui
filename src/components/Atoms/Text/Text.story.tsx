import React from "react";
import { storiesOf } from "@storybook/react-native";
import Text from "./Text";
import Screen from "../Screen/Screen";
import Box from "../Box/Box";

storiesOf("Typography", module)
  .addDecorator((getStory) => <Screen>{getStory()}</Screen>)
  .add("All", () => (
    <Box flex={1}>
      <Text variant="h1">Heading 1</Text>
      <Text variant="h2">Heading 2</Text>
      <Text variant="h3">Heading 3</Text>
      <Text variant="h4">Heading 4</Text>
      <Text variant="h5">Heading 5</Text>
      <Text variant="h6">Heading 6</Text>
      <Text variant="p1">Paragraph 1</Text>
      <Text variant="p2">Paragraph 2</Text>
      <Text variant="p3">Paragraph 3</Text>
      <Text variant="p4">Paragraph 4</Text>
      <Text variant="label">Label</Text>
      <Text variant="caption">Caption</Text>
      <Box backgroundColor="primary.500">
        <Text variant="btn1">Button 1</Text>
      </Box>
      <Box backgroundColor="primary.500">
        <Text variant="btn2">Button 2</Text>
      </Box>
      <Box backgroundColor="primary.500">
        <Text variant="btn3">Button 3</Text>
      </Box>
      <Box backgroundColor="primary.500">
        <Text variant="btn4">Button 4</Text>
      </Box>
    </Box>
  ));
