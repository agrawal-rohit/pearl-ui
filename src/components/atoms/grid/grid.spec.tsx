import React from "react";
import { render } from "@testing-library/react-native";
import Grid from "./grid";
import { ThemeProvider } from "../../../theme/src/theme-context";
import Box from "../box/box";
import Text from "../text/text";

jest.useFakeTimers();

describe("Atoms/Grid", () => {
  it("passes the snapshot test with default props", () => {
    const tree = render(
      <ThemeProvider>
        <Grid>
          <Box>
            <Text>Test</Text>
          </Box>
        </Grid>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("passes the snapshot test with custom numCols", () => {
    const tree = render(
      <ThemeProvider>
        <Grid numCols={3}>
          <Box>
            <Text>Test</Text>
          </Box>
          <Box>
            <Text>Test</Text>
          </Box>
          <Box>
            <Text>Test</Text>
          </Box>
        </Grid>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("passes the snapshot test with custom spacing", () => {
    const tree = render(
      <ThemeProvider>
        <Grid spacing="5">
          <Box>
            <Text>Test</Text>
          </Box>
        </Grid>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("passes the snapshot test with custom spacingX", () => {
    const tree = render(
      <ThemeProvider>
        <Grid spacingX="2">
          <Box>
            <Text>Test</Text>
          </Box>
        </Grid>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("passes the snapshot test with custom spacingY", () => {
    const tree = render(
      <ThemeProvider>
        <Grid spacingY="2">
          <Box>
            <Text>Test</Text>
          </Box>
        </Grid>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
