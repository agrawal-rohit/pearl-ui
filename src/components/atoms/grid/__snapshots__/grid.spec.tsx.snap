// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms/Grid passes the snapshot test 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    animate={{}}
    animatedStyle={
      {
        "value": {},
      }
    }
    collapsable={false}
    style={{}}
    transition={
      {
        "dampingRatio": 1,
        "duration": 100,
        "type": "spring",
      }
    }
  />
</View>
`;
